share/doc/wake/tutorial.md

```diff
@@ -29,112 +29,224 @@ The verbose output is of the form `expression: type = value`. The above
+would be `f(x, y)`. So compileC is being run on four arguments.
 
-The arguments to compileC are:
-  1. the build variant as a String
-  2. a list of additional compiler options
-  3. a list of legal input files (more on this later)
-  4. the name of the file to compile
+    wake --in gcc_wake -vx compileC
 
-The output of compileC is the `Path` of the object file produced.
+Notice that the type of compileC is `(variant: String) => (extraFlags: List String) => (headers: List Path) => (cfile: Path) => Result (List Path) Error`.
+This should be read as "a function that takes a `String` named `variant`,
+then a `List` of `Strings` named `extraFlags`, then a `List` of `Paths` named
+`headers`, another Path named `cfile`, and finally returns a `List Path` if
+nothing failed but an `Error` if something did."
+
+Indeed, we can see in our use of compileC, we passed a `String` for the
+first argument and `main` -- which is indeed a `Path` -- for the last argument.
+The second and third arguments are `Lists`; `Nil` is the empty list and `(x, y, Nil)`
+is a list with two elements x and y. We'll look at the returned value later.
+
+Back to the `require` line, `Pass` is a constructor to the `Result` type which
+takes a single argument. Unlike compileC, `source` returns a `Result Path Error`
+(i.e. a single `Path` instead of multiple). If wake can't find `main.cpp` --
+likely because it hasn't been added to git -- then we don't have any need or
+desire to try compiling a non-existent file, so if the `require` gets passed a
+`Fail` instead of a `Pass` (the second constructor to `Result`; we'll cover that
+more later) we break out of `build` early and return that failure.
+
+Unlike `compileC`, the `source` function returns a `Result Path Error` so its
+`Pass` constructor and destructor (and thus `main`) only wrap a single `Path`.
+This is also where the git commands come into play, as `source` only searches
+the git index to find files. Don't forget to add anything you create!
 
 ## Compiling and linking two CPP files
 
     echo 'int helper() { return 42; }' > help.cpp
     git add help.cpp
     cat > tutorial.wake <<'EOF'
+    from wake import _
+    from gcc_wake import compileC linkO
     def variant = "native-cpp11-release"
     def build _ =
-      def main = compileC variant ("-I.", Nil) Nil (source "main.cpp")
-      def help = compileC variant ("-I.", Nil) Nil (source "help.cpp")
-      linkO variant ("-lm", Nil) (main, help, Nil) "tutorial"
+        def mainResult =
+            require Pass mainSrc =
+                source "main.cpp"
+            compileC variant ("-I.", Nil) Nil mainSrc
+        def helpResult =
+            require Pass helpSrc =
+                source "help.cpp"
+            compileC variant ("-I.", Nil) Nil helpSrc
+        def tutorialResult =
+            require Pass main = mainResult
+            require Pass help = helpResult
+            linkO variant ("-lm", Nil) (main, help, Nil) "tutorial"
+        tutorialResult
```

Comment:
I'm still so torn on weather we should write it like this or not:

```
      require Pass main = compileC variant ("-I.", Nil) Nil (source "main.cpp")
      require Pass help = compileC variant ("-I.", Nil) Nil (source "help.cpp")
      linkO variant ("-lm", Nil) (main, help, Nil) "tutorial"
```

I know its wrong but its way less confusing...I'm not sure which does more damage. Unfortunately explaining *why* this is wrong is more confusing than just showing it the right way.

Increasingly I think wake should just automatically de-serialize these lines to avoid this issue entirely.